cpu/saml21/cpu.c

```diff
@@ -0,0 +1,58 @@
+/*
+ * Copyright (C) 2015 Kaspar Schleiser <kaspar@schleiser.de>
+ *               2015 FreshTemp, LLC.
+ *
+ * This file is subject to the terms and conditions of the GNU Lesser
+ * General Public License v2.1. See the file LICENSE in the top level
+ * directory for more details.
+ */
+
+/**
+ * @ingroup     cpu_saml21
+ * @{
+ *
+ * @file        cpu.c
+ * @brief       Implementation of the CPU initialization for Atmel SAML21 MCUs
+ *
+ * @author      Kaspar Schleiser <kaspar@schleiser.de>
+ * @}
+ */
+
+#include "cpu.h"
+
+/**
+ * @brief Initialize the CPU, set IRQ priorities, clocks
+ */
+void cpu_init(void)
+{
+    /* disable the watchdog timer */
+    WDT->CTRLA.bit.ENABLE = 0;
+
+    /* set pendSV interrupt to lowest possible priority */
+    NVIC_SetPriority(PendSV_IRQn, 0xff);
+
+    /* turn on MCLK */
+    MCLK->APBAMASK.reg |= MCLK_APBAMASK_GCLK;
+
+    /*OSCCTRL->OSC16MCTRL.reg |= OSCCTRL_OSC16MCTRL_FSEL(OSCCTRL_OSC16MCTRL_FSEL_16) |
+            OSCCTRL_OSC16MCTRL_RUNSTDBY;
+            ;
+*/
+
+    /* Software reset the GCLK module to ensure it is re-initialized correctly */
+    GCLK->CTRLA.reg = GCLK_CTRLA_SWRST;
+    while (GCLK->CTRLA.reg & GCLK_CTRLA_SWRST);
+    while (GCLK->SYNCBUSY.reg & GCLK_SYNCBUSY_SWRST);
+
+    /* Select the correct generator */
+    *((uint8_t*)&GCLK->GENCTRL[0].reg) = 0;
+    while (GCLK->SYNCBUSY.reg & GCLK_SYNCBUSY_SWRST);
+    GCLK->GENCTRL[0].reg = (
+            GCLK_GENCTRL_GENEN          /* enable gclk */
+            | GCLK_GENCTRL_SRC_OSC16M   /* select 16mhz osc */
```

Comment:
actually, I would strongly opt for initializing the CPU to its maximum possible frequency (48MHz?) for consistency reasons (we do this for every board in RIOT). In my opinion the paradigm should be: start out with the max clock and then reduce this based on your application needs if you really need to save power...